lib/wallet/bench/api-bench.hs

```diff
@@ -0,0 +1,316 @@
+-------------------------------------------------------------------------------}
+main :: IO ()
+main = do
+    Sys.getArgs >>= \case
+        [] -> do
+            sayErr "expected 1 argument: directory with wallet database files"
+            Sys.exitFailure
+        (dir: _) -> benchmarkApi dir
+
+benchmarkApi :: FilePath -> IO ()
+benchmarkApi dir = do
+    (_, tr) <- initBenchmarkLogging "wallet" Notice
+    runBenchmarks
+        [ benchmarkWallets "sequential" dir tr (Proxy @('Testnet 0)) benchmarksSeq
+        ]
+
+{-------------------------------------------------------------------------------
+    API benchmarks
+-------------------------------------------------------------------------------}
+data WalletOverview = WalletOverview
+    { utxo :: UTxOStatistics
+    , addresses :: Word
+    , transactions :: Word
+    } deriving (Show, Generic)
+
+instance Buildable WalletOverview where
+    build WalletOverview{utxo,addresses,transactions} =
+        blockListF' "" id
+            [ nameF "number of addresses" (build addresses)
+            , nameF "number of transactions" (build transactions)
+            , build utxo
+            ]
+
+instance ToJSON WalletOverview where
+    toJSON WalletOverview{utxo,addresses,transactions} = Aeson.object
+        [ "utxo" .= jsonFromUTxOStatistics utxo
+        , "addresses" .= addresses
+        , "transactions" .= transactions
+        ]
+
+jsonFromUTxOStatistics :: UTxOStatistics -> Aeson.Value
+jsonFromUTxOStatistics UTxOStatistics{histogram,allStakes,boundType} =
+    Aeson.object
+        [ "total" .=
+            (Quantity (fromIntegral allStakes) :: Quantity "lovelace" Natural)
+        , "scale" .= genericToJSON Aeson.defaultOptions boundType
+        , "distribution" .=
+            Map.fromList (map (\(HistogramBar k v)-> (k,v)) histogram)
+        ]
+
+data BenchSeqResults = BenchSeqResults
+    { benchName :: Text
+    , readWalletTime :: Time
+    , listAddressesTime :: Time
+    , listTransactionsTime :: Time
+    , walletOverview :: WalletOverview
+    } deriving (Show, Generic)
+
+instance Buildable BenchSeqResults where
+    build = genericF
+
+instance ToJSON BenchSeqResults where
+    toJSON = genericToJSON Aeson.defaultOptions
+
+data BenchmarkConfig (n :: NetworkDiscriminant) s k = BenchmarkConfig
+    { benchmarkName :: Text
+    , networkId :: Proxy n
+    , ctx :: MockWalletLayer IO s k
+    , wid :: WalletId
+    }
+
+benchmarksSeq
+    :: forall (n :: NetworkDiscriminant) s k.
+        ( s ~ SeqState n k
+        , k ~ ShelleyKey
+        , Typeable n
+        , PaymentAddress n k 'CredFromKeyK
+        , NetworkDiscriminantVal n
+        )
+    => BenchmarkConfig n s k
+    -> IO BenchSeqResults
+benchmarksSeq BenchmarkConfig{benchmarkName,ctx,wid} = do
+    ((cp, pending), readWalletTime) <- bench "read wallet" $ do
+        (cp, _, pending) <- unsafeRunExceptT $ W.readWallet @_ @s @k ctx wid
+        pure (cp, pending)
+
+    (utxo, _) <- bench "utxo statistics" $
```

Comment:
From the run on https://buildkite.com/input-output-hk/cardano-wallet-nightly/builds/1722#0186ea82-85b2-49c5-a140-c68ee19d07f8 we see we have those:
 ``` 
  readWalletTime: 548.0 Î¼s
  listAddressesTime: 820.7 Î¼s
  listTransactionsTime: 237.1 ms
```
let's add few more if possible:
 - https://input-output-hk.github.io/cardano-wallet/api/edge/#operation/getUTxOsStatistics
 - https://input-output-hk.github.io/cardano-wallet/api/edge/#operation/getWalletUtxoSnapshot
 - https://input-output-hk.github.io/cardano-wallet/api/edge/#operation/listAssets
 - https://input-output-hk.github.io/cardano-wallet/api/edge/#operation/listWallets (if that makes sense)
 - https://input-output-hk.github.io/cardano-wallet/api/edge/#operation/createShelleyWalletMigrationPlan (if that makes sense)
(perhaps it would be good to have something that's doing coin selection, so one of or all?:)
 - https://input-output-hk.github.io/cardano-wallet/api/edge/#operation/selectCoins 
 - https://input-output-hk.github.io/cardano-wallet/api/edge/#operation/postTransactionFee
 - https://input-output-hk.github.io/cardano-wallet/api/edge/#operation/constructTransaction

I think that would make fairly broad set that makes sense to measure on DB (what do you think?). I'll provide some reasonable databases for Byron random and Shelley seq. 1 for each should be enough.

Reply:
https://input-output-hk.github.io/cardano-wallet/api/edge/#operation/getDelegationFee would be worth adding as well. (ATM, the only fee that's using balanceTx, see: https://github.com/input-output-hk/cardano-wallet/pull/3787#issuecomment-1477831332)